### PR TITLE
MLDB-1738 tensorflow builtin functions

### PR DESCRIPTION
**Pull request by jeremybarnes**
Adds the ability to call a Tensorflow function as a builtin by appending tf_ to its name.

### CODE REVIEW

**Comment by jeremybarnes** [comment]
Ignore the test failure which is simple to fix

**Comment by jeremybarnes** [comment]
Ignore the test failure which is simple to fix

**Comment by mailletf** [comment]
Are there any stability implications for this PR on existing functionality?

**Comment by jeremybarnes** [comment]
There shouldn't be.  The only code paths changed are in the processing of embeddings but they were essentially bug fixes.

**Comment by mailletf** [comment]
Ok. Just to be on the safe side, since I want to do a release Tuesday so let's not merge into master until then. That'll give us more time to shake out bugs if any arise.

**Comment by jeremybarnes** [comment]
OK, 10-4 on that

**Comment by jeremybarnes** [comment]
Test failure is once again spurious.

File: sql/expression_value.cc [hunk 1]
```diff
@@ -1896,62 +1973,59 @@ ExpressionValue(const std::vector<double> & values,
...
 {
-    std::shared_ptr<double> vals(new double[values.size()],
-                                    [] (double * p) { delete[] p; });
-    std::copy(std::make_move_iterator(values.begin()),
-              std::make_move_iterator(values.end()),
-              vals.get());
+    // This avoids needing to reallocate... it essentially allows us to create
+    // a shared_ptr that owns the storage of a vector
+    auto movedVals = std::make_shared<std::vector<double> >(std::move(values));
+    std::shared_ptr<double> vals(movedVals, movedVals->data());
```
Comment (FinchPowers) [inline sql/expression_value.cc]: Culture personnelle: Un double initialisé avec 2 paramètres ça fait quoi?

File: tensorflow/tensorflow_plugin.cc [hunk 2]
```diff
@@ -838,12 +742,190 @@ struct TensorflowGraph: public Function {
...
+                                     len, ST_INT64);
+                return result;
+            }
+            default:
+                break;
+            }
+        }
+
+        cerr << "val = " << jsonEncode(val) << endl;
+        cerr << "type = " << type << endl;
```
Comment (FinchPowers) [inline tensorflow/tensorflow_plugin.cc]: Est-ce qu'on veut garder ça? Si oui on veut probablement utiliser le logger.

File: tensorflow/tensorflow_plugin.cc [hunk 3]
```diff
@@ -1179,12 +1309,646 @@ struct TensorflowGraph: public Function {
...
+            throw HttpReturnException(400, "Unable to obtain TensorFlow operator '"
+                                      + functionConfig.op + "': "
+                                      + status.error_message());
+        }
+
+        // Create a graph containing just the op
+        std::unique_ptr<tensorflow::GraphDef>
+            graph(new tensorflow::GraphDef());
+
+        for (auto & input: op->input_arg()) {
```
Comment (FinchPowers) [inline tensorflow/tensorflow_plugin.cc]: const auto?

File: tensorflow/tensorflow_plugin.cc [hunk 4]
```diff
@@ -1179,12 +1309,646 @@ struct TensorflowGraph: public Function {
...
+        node->set_device("/cpu:0");
+
+        auto * nattr = node->mutable_attr();
+
+        for (auto & attr: functionConfig.attr) {
+            (*nattr)[attr.first.rawString()] = attr.second;
+        }
+
+        // Create a node for each input
+        for (auto & input: op->input_arg()) {
```
Comment (FinchPowers) [inline tensorflow/tensorflow_plugin.cc]: const?

File: tensorflow/tensorflow_plugin.cc [hunk 5]
```diff
@@ -1179,12 +1309,646 @@ struct TensorflowGraph: public Function {
...
+        context.writeHtml("</ul>");
+    }
+
+    template<typename T>
+    void writeAttrListT(MacroContext & context,
+                       const ::google::protobuf::RepeatedPtrField<T>& val)
+    {
+        context.writeHtml("<ul>");
+
+        for (auto & v: val) {
```
Comment (FinchPowers) [inline tensorflow/tensorflow_plugin.cc]: const?

File: tensorflow/tensorflow_plugin.cc [hunk 6]
```diff
@@ -1179,12 +1309,646 @@ struct TensorflowGraph: public Function {
...
+            context.writeHtml("</code> Operation</h2>");
+            context.writeHtml("<h3>Description</h3>");
+            context.writeMarkdown(opDef->description());
+
+            if (opDef->attr_size()) {
+                context.writeHtml("<h3>Attributes</h3>");
+                context.writeHtml("<table><tr><th>Attribute</th><th>Type</th>"
+                                  "<th>Default</th><th>Constraints</th>"
+                                  "<th>Description</th></tr>");
+                for (auto & a: opDef->attr()) {
```
Comment (FinchPowers) [inline tensorflow/tensorflow_plugin.cc]: const?

**Comment by FinchPowers** [comment]
+1

File: sql/expression_value.cc (hunk 1 above)
Comment (jeremybarnes) [inline sql/expression_value.cc]: It's a shared pointer, intialized with contents controlled by the second, and lifetime controlled by the first.  Look at the shared_ptr constructor documentation.  It's really neat for, eg, passing a shared_ptr to a structure field but with the shared_ptr owning the entire structure.

**Pull request merged**